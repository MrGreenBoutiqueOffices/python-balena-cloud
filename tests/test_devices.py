"""Test the devices model for Balena Cloud."""

# pylint: disable=too-many-arguments,too-many-positional-arguments
from __future__ import annotations

from typing import TYPE_CHECKING, Any

import pytest
from aresponses import ResponsesMockServer
from syrupy.assertion import SnapshotAssertion

from . import load_fixtures

if TYPE_CHECKING:
    from balena_cloud import BalenaCloud


@pytest.mark.parametrize(
    ("param_key", "param_value", "endpoint"),
    [
        ("device_id", 1, "/v7/device(1)"),
        ("device_uuid", "test-uuid", "/v7/device(uuid='test-uuid')"),
    ],
)
async def test_get_device(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    balena_cloud_client: BalenaCloud,
    param_key: str,
    param_value: Any,
    endpoint: str,
) -> None:
    """Test the get_device method."""
    aresponses.add(
        "api.balena-cloud.com",
        endpoint,
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("devices/device.json"),
        ),
    )
    # Get the device based on the parameter key
    device = None
    if param_key == "device_id":
        device = await balena_cloud_client.device.get(device_id=param_value)
    elif param_key == "device_uuid":
        device = await balena_cloud_client.device.get(device_uuid=param_value)
    assert device == snapshot


async def test_update_device(
    aresponses: ResponsesMockServer,
    balena_cloud_client: BalenaCloud,
) -> None:
    """Test the update_device method."""
    aresponses.add(
        "api.balena-cloud.com",
        "/v7/device(1)",
        "PATCH",
    )
    await balena_cloud_client.device.update(
        device_id=1, data={"device_name": "Test Device"}
    )


async def test_remove_device(
    aresponses: ResponsesMockServer,
    balena_cloud_client: BalenaCloud,
) -> None:
    """Test the remove_device method."""
    aresponses.add(
        "api.balena-cloud.com",
        "/v7/device(1)",
        "DELETE",
    )
    await balena_cloud_client.device.remove(device_id=1)


@pytest.mark.parametrize(
    ("param_key", "param_value"),
    [
        ("device_id", 1),
        ("device_uuid", "test-uuid"),
    ],
)
async def test_get_device_tags(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    balena_cloud_client: BalenaCloud,
    param_key: str,
    param_value: Any,
) -> None:
    """Test the get_device_tags method."""
    aresponses.add(
        "api.balena-cloud.com",
        "/v7/device_tag",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("devices/device_tags.json"),
        ),
    )
    # Get the device tags based on the parameter key
    tags = None
    if param_key == "device_id":
        tags = await balena_cloud_client.device_tag.get_all(device_id=param_value)
    elif param_key == "device_uuid":
        tags = await balena_cloud_client.device_tag.get_all(device_uuid=param_value)
    assert tags == snapshot


async def test_add_device_tag(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    balena_cloud_client: BalenaCloud,
) -> None:
    """Test the add_device_tag method."""
    aresponses.add(
        "api.balena-cloud.com",
        "/v7/device_tag",
        "POST",
        aresponses.Response(
            status=201,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("devices/post_device_tag.json"),
        ),
    )
    tag = await balena_cloud_client.device_tag.add(
        device_id=1, key="test_key", value="test_value"
    )
    assert tag == snapshot


async def test_update_device_tag(
    aresponses: ResponsesMockServer,
    balena_cloud_client: BalenaCloud,
) -> None:
    """Test the update_device_tag method."""
    aresponses.add(
        "api.balena-cloud.com",
        "/v7/device_tag(device=1,tag_key='test_key')",
        "PATCH",
    )
    await balena_cloud_client.device_tag.update(
        device_id=1, key="test_key", value="test_value"
    )


async def test_remove_device_tag(
    aresponses: ResponsesMockServer,
    balena_cloud_client: BalenaCloud,
) -> None:
    """Test the remove_device_tag method."""
    aresponses.add(
        "api.balena-cloud.com",
        "/v7/device_tag(1)",
        "DELETE",
    )
    await balena_cloud_client.device_tag.remove(tag_id=1)


@pytest.mark.parametrize(
    ("param_key", "param_value"),
    [
        ("device_id", 1),
        ("device_uuid", "test-uuid"),
    ],
)
async def test_get_device_variables(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    balena_cloud_client: BalenaCloud,
    param_key: str,
    param_value: Any,
) -> None:
    """Test the get_device_variables method."""
    aresponses.add(
        "api.balena-cloud.com",
        "/v7/device_environment_variable",
        "GET",
        aresponses.Response(
            status=200,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("devices/device_variables.json"),
        ),
    )
    # Get the device variables based on the parameter key
    variables = None
    if param_key == "device_id":
        variables = await balena_cloud_client.device_variable.get_all(
            device_id=param_value
        )
    elif param_key == "device_uuid":
        variables = await balena_cloud_client.device_variable.get_all(
            device_uuid=param_value
        )
    assert variables == snapshot


async def test_add_device_variable(
    aresponses: ResponsesMockServer,
    snapshot: SnapshotAssertion,
    balena_cloud_client: BalenaCloud,
) -> None:
    """Test the add_device_variable method."""
    aresponses.add(
        "api.balena-cloud.com",
        "/v7/device_environment_variable",
        "POST",
        aresponses.Response(
            status=201,
            headers={"Content-Type": "application/json"},
            text=load_fixtures("devices/post_device_variable.json"),
        ),
    )
    variable = await balena_cloud_client.device_variable.add(
        device_id=1, name="test_name", value="test_value"
    )
    assert variable == snapshot


async def test_update_device_variable(
    aresponses: ResponsesMockServer,
    balena_cloud_client: BalenaCloud,
) -> None:
    """Test the update_device_variable method."""
    aresponses.add(
        "api.balena-cloud.com",
        "/v7/device_environment_variable(1)",
        "PATCH",
    )
    await balena_cloud_client.device_variable.update(variable_id=1, value="test_value")


async def test_remove_device_variable(
    aresponses: ResponsesMockServer,
    balena_cloud_client: BalenaCloud,
) -> None:
    """Test the remove_device_variable method."""
    aresponses.add(
        "api.balena-cloud.com",
        "/v7/device_environment_variable(1)",
        "DELETE",
    )
    await balena_cloud_client.device_variable.remove(variable_id=1)
