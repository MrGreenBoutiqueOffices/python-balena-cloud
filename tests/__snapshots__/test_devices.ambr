# serializer version: 1
# name: test_add_device_tag
  Tag(id=1, key='tag_1', value='value_1')
# ---
# name: test_add_device_variable
  EnvironmentVariable(id=1, name='VARIABLE_1', value='1234567890', created_at=datetime.datetime(2024, 1, 1, 0, 0, tzinfo=datetime.timezone.utc))
# ---
# name: test_get_device[device_id-1-/v7/device(1)]
  Device(id=1, name='Device_1', status='Idle', uuid='00000000000000000000000000000001', is_online=True, is_web_accessible=True, is_undervolted=False, latitude=0.0, longitude=0.0)
# ---
# name: test_get_device[device_uuid-test-uuid-/v7/device(uuid='test-uuid')]
  Device(id=1, name='Device_1', status='Idle', uuid='00000000000000000000000000000001', is_online=True, is_web_accessible=True, is_undervolted=False, latitude=0.0, longitude=0.0)
# ---
# name: test_get_device_tag
  Tag(id=1, key='tag_1', value='value_1')
# ---
# name: test_get_device_tags[device_id-1]
  list([
    Tag(id=1, key='tag_1', value='value_1'),
    Tag(id=2, key='tag_2', value='value_2'),
    Tag(id=3, key='tag_3', value='value_3'),
  ])
# ---
# name: test_get_device_tags[device_uuid-test-uuid]
  list([
    Tag(id=1, key='tag_1', value='value_1'),
    Tag(id=2, key='tag_2', value='value_2'),
    Tag(id=3, key='tag_3', value='value_3'),
  ])
# ---
# name: test_get_device_variable
  EnvironmentVariable(id=1, name='VARIABLE_1', value='1234567890', created_at=datetime.datetime(2024, 1, 1, 0, 0, tzinfo=datetime.timezone.utc))
# ---
# name: test_get_device_variables[device_id-1]
  list([
    EnvironmentVariable(id=1, name='VARIABLE_1', value='1234567890', created_at=datetime.datetime(2024, 1, 1, 0, 0, tzinfo=datetime.timezone.utc)),
    EnvironmentVariable(id=2, name='VARIABLE_2', value='TEST_1', created_at=datetime.datetime(2024, 1, 2, 0, 0, tzinfo=datetime.timezone.utc)),
  ])
# ---
# name: test_get_device_variables[device_uuid-test-uuid]
  list([
    EnvironmentVariable(id=1, name='VARIABLE_1', value='1234567890', created_at=datetime.datetime(2024, 1, 1, 0, 0, tzinfo=datetime.timezone.utc)),
    EnvironmentVariable(id=2, name='VARIABLE_2', value='TEST_1', created_at=datetime.datetime(2024, 1, 2, 0, 0, tzinfo=datetime.timezone.utc)),
  ])
# ---
